### PR TITLE
Do not access join layer from QgsVectorLayerFeatureIterator

### LINKED ISSUE #38551: Infinite hang with a layer style based on a joined column
**Describe the bug**

When displaying a layer joined to another layer and styling it based on a joined column, the rendering performance drops a lot, sometimes resulting in an infinite "Not responding" Window.

**How to Reproduce**

1. Download the demo project [here](https://sitn.ne.ch/web/qgis/test_join.zip).
2. Unzip and open `join.qgz`, you should see red rectangles.
3. Zoom in / out -> everything is ok. The active layer has a simple style.
4. Activate _joined_polygons_categorized_ layer which has a style based on the joined layer
5. Zoom out -> progress bar in the status bar takes some time.
6. Do some zoom in / out quick, progress bar stays "progressing", trying to do exit QGIS will result in a "Not responding" window.

**QGIS and OS versions**

Version de QGIS | 3.14.15-Pi | Révision du code | d5114d2cfa
-- | -- | -- | --
Compilé avec Qt | 5.11.2 | Utilisant Qt | 5.11.2
Compilé avec GDAL/OGR | 3.0.4 | Utilisé avec GDAL/OGR | 3.0.4
Compilé avec GEOS | 3.8.1-CAPI-1.13.3 | Utilisé avec GEOS | 3.8.1-CAPI-1.13.3
Compilé avec SQLite | 3.29.0 | Fonctionne avec SQLite | 3.29.0
Version du client PostgreSQL | 11.5 | Version de SpatiaLite | 4.3.0
Version de QWT | 6.1.3 | Version de QScintilla2 | 2.10.8
Compilé avec PROJ | 6.3.2 | Fonctionne avec PROJ | Rel. 6.3.2, May 1st, 2020
Version de l'OS | Windows 10 (10.0)
Extensions Python actives | changeDataSource; searchlayers; MetaSearch; processing

**Additional context**

Same behavior if the layers are on PostGIS.

### PR DESCRIPTION
**Pull request by nyalldawson**
This is not thread safe at all - we cannot access a layer from
an iterator, as iterators may be running on background threads.

Instead use a thread safe approach of storing a QgsVectorLayerFeatureSource
and using that instead

Fixes #38551
\

### CODE REVIEW

**Comment by nyalldawson** [comment]
@wonder-sk I'd love your eyes on this one, if you can spare a moment!

File: src/core/vector/qgsvectorlayerfeatureiterator.cpp [hunk 1]
```diff
@@ -1045,11 +1049,13 @@ void QgsVectorLayerFeatureIterator::FetchJoinInfo::addJoinedAttributesCached( Qg
 
 void QgsVectorLayerFeatureIterator::FetchJoinInfo::addJoinedAttributesDirect( QgsFeature &f, const QVariant &joinValue ) const
 {
+#if 0 // this is not thread safe -- we cannot access the layer here as this will be called from non-main threads.
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayerfeatureiterator.cpp:1052]: Ping @elpaso -- I've had to disable this shortcut -- there's no easy way to make this thread safe. The hasFeatures method would need to be added to QgsVectorLayerFeatureSource and all the various provider's QgsAbstractFeatureSource subclasses, but that's not a trivial task to do.

Comment (m-kuhn) [inline src/core/vector/qgsvectorlayerfeatureiterator.cpp:1052]: Why keep it?

Comment (nyalldawson) [reply src/core/vector/qgsvectorlayerfeatureiterator.cpp:1052]: it'd be a nice optimisation to revive at some stage, if the required api is added

**Comment by nyalldawson** [comment]
Unrelated test failure (ping @pblottiere -- something has started failing in the ogc tests for all PRs/branches. Looks like the test itself is broken, not related to a QGIS change)

**Comment by pblottiere** [comment]
@nyalldawson the underlying OGC docker image  https://hub.docker.com/r/ogccite/ets-wms13 has been updated 2 days ago so it's probably related. I'll take a look tomorrow morning.

**Comment by nyalldawson** [comment]
Review ping?

**Review by 3nids — APPROVED** [review]

**Pull request merged**